ocserv: T3896: add CLI options to configure ocserv config-per-user/group

PeppyH opened the pull request:
Adds CLI configurations under VPN - OpenConnect to facilitate per user/group vpn session configurations. Validation has been added to restrict config-per-group to be exclusive to OpenConnect RADIUS authentication as the config file is looked up based on a RADIUS response attribute - as well as sanity check that the necessary configs are configured when not disabled.

<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
Adds:
* New interface include `openconnect-identity-based-config.xml.i`

Changes:
* Updates interface `vpn-openconnect.xml.in` to include `openconnect-identity-based-config.xml.i` as child node of authentication
* Updates template file `ocserv_config.j2` to add `config-per-user` or `config-per-group` where applicable based on CLI configuration
* Updates `vpn_openconnect.py` to add config validation in the `verify` stage

I wasn't sure what to name the configuration node - `config-per-user-or-group` seems too verbose and I think the help node does the job of explaining what the command is responsible for configuring, however I don't have a strong opinion either way - happy for you guys to change it.

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- All submitted PRs must be linked to a Task on Phabricator. -->
* https://phabricator.vyos.net/T3896

## Component(s) name
ocserv
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->

## Proposed changes
<!--- Describe your changes in detail -->
Adds ocserv configuration via VyOS CLI to specify per user/group configurations. When identity-based-config is set to user mode it can be used with the local authentication method as it matches the username to match to a config file. However if configured for group mode the openconnect authentication mode must be set to radius - `vpn_openconnect.py` validates that in its verify function.

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->
### Prerequisite
If you don't already have a valid ssl configuration for your openconnect vpn follow [these steps](https://docs.vyos.io/en/latest/configuration/vpn/openconnect.html?highlight=openconnect#ssl-certificates-generation) and setup a radius server for authentication

```
set vpn openconnect authentication mode radius
set vpn openconnect authentication radius server 10.0.95.49 port 1812
set vpn openconnect authentication radius server 10.0.95.49 key your_radius_secret_here
```

### Testing Generated Configs
```
sudo mkdir -p /config/auth/ocserv/config-per-user
sudo touch /config/auth/ocserv/default-user.conf

set vpn openconnect authentication identity-based-config mode user
set vpn openconnect authentication identity-based-config directory /config/auth/ocserv/config-per-user
set vpn openconnect authentication identity-based-config default-config /config/auth/ocserv/default-user.conf
commit

sudo cat /run/ocserv/ocserv.conf
```

The output ocserv.conf should include two lines as per below:
```
config-per-user = /config/auth/ocserv/config-per-user
default-user-config = /config/auth/ocserv/default-user.conf
```

### Testing by connecting to the VPN
Create a config file for the user you will be testing with and enter some data, e.g.
```
nano /config/auth/ocserv/config-per-user/{yourusername}
## enter the following:
dns = 8.8.4.4
```
Connect to the VPN with openconnect e.g. and observe your DNS change
```
openconnect -u {yourusername} https://{your openconnect vpn endpoint}
```

You can repeat these steps switching
* vpn authentication configuration mode to `local`
* change `identity-based-config` mode to group and have your RADIUS server reply with a Class attribute matching a file in the configured directory.
## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

Commit:
ocserv: T3896: add CLI options to configure ocserv config-per-user/group

Adds CLI configurations under VPN - OpenConnect to facilitate per user/group vpn session configurations. Validation has been added to restrict config-per-group to be exclusive to OpenConnect RADIUS authentication as the config file is looked up based on a RADIUS response attribute - as well as sanity check that the necessary configs are configured when not disabled.

dmbaturin commented on interface-definitions/include/openconnect-config-per-x.xml.i:
```diff
@@ -0,0 +1,54 @@
+<!-- include start from openconnect-config-per-x.xml.i -->
+<node name="config-per-x">
```
I get your idea but I believe we can find a better name for this node than `config-per-x`. Like `external-config` maybe.

dmbaturin commented on interface-definitions/include/openconnect-config-per-x.xml.i:
```diff
@@ -0,0 +1,54 @@
+<!-- include start from openconnect-config-per-x.xml.i -->
+<node name="config-per-x">
+    <properties>
+        <help>Configures ocserv to search the configured directory for a config file matching the Group name or username</help>
```
The fact that the daemon is named "ocserv" is an implementation detail. I believe in user-visible parts, it should always be "OpenConnect".

dmbaturin commented on interface-definitions/include/openconnect-config-per-x.xml.i:
```diff
@@ -0,0 +1,54 @@
+    <children>
+        <leafNode name="mode">
+            <properties>
+                <help>ocserv will ignore these configs if groupconfig is enabled TODO: explain this in better detail - need to make it clear this is in reference to the groupconfig vyos config that lives up one level from here</help>
```
Before there's a final version of the wording, perhaps it should be a draft PR. ;)

PeppyH replied:
AH, so in my workflow I add a bunch of these TODO's and work through them one by one and when I commit changes I diff them and review them and search for 'TODO' to make sure I got them all.

I must have missed this one - I didn't intend to submit a PR with this in it, I'll get that fixed up.

dmbaturin commented on interface-definitions/include/openconnect-config-per-x.xml.i:
```diff
@@ -0,0 +1,54 @@
+                <constraint>
+                    <regex>(user|group)</regex>
+                </constraint>
+                <constraintErrorMessage>Invalid config-per-x. Must be one of: user,  group </constraintErrorMessage>
```
This is a rather unhelpful message because it doesn't mention that the offending node is `mode`, not the upper level `config-per-x`.

dmbaturin reviewed (changes requested):
I believe the CLI can be improved.

Commit:
ocserv: T3896: CLI help and description rewording

PeppyH commented:
Thanks for the feedback, I've pushed new changes. I can squash these into a single commit if you would like prior to merging.

With regards to the node name i'd like to get some other opinions on the matter before I change it, but ultimately I agree `config-per-x` can be improved. Some other possibilities:
* `per-session-config`
* `per-user-or-group-config`
* `config-per-user-or-group`
* `identity-based-config`

sever-sever commented:
For now it is  still `config-per-x/config_per_x`

dmbaturin commented:
> For now it is still `config-per-x/config_per_x`

Ah, wait, I missed it (there are certainly many improvements, though). I'm inclined towards `identity-based-config`, what do you think?

PeppyH commented:
> > For now it is still `config-per-x/config_per_x`
> 
> Ah, wait, I missed it (there are certainly many improvements, though). I'm inclined towards `identity-based-config`, what do you think?

I'm leaning towards this option as well, i'm going to push new commits which will change the node name.

Update: I've updated the original comment to account for the new node name change

Commit:
ocserv: T3896: refactor: change ocserv config-per-x node name

Changes the node name from config-per-x to identity-based-config, as a result the j2 templates and vpn_openeconnect.py has been refactored to update the node name when accessing it's child nodes.
